lib/SIL/DynamicCasts.cpp

```diff
@@ -390,6 +390,21 @@ swift::classifyDynamicCast(ModuleDecl *M,
             sourceMetatype.isAnyExistentialType())
       return DynamicCastFeasibility::WillSucceed;
 
+    // If the source and target are the same existential type, but the source is
+    // P.Protocol and the dest is P.Type, then we need to consider whether the
+    // protocol is self-conforming.
+    // The only cases where a protocol self-conforms are objc protocols, but
+    // we're going to expect P.Type to hold a class object. And this case
+    // doesn't matter since for a self-conforming protocol type there can't be
+    // any type-level methods.
+    // Thus we consider this kind of cast to always fail.
+    if (source == target &&
+        (isa<ProtocolType>(source) || isa<ProtocolCompositionType>(source)) &&
+        isa<MetatypeType>(sourceMetatype) &&
+        isa<ExistentialMetatypeType>(targetMetatype)) {
+      return DynamicCastFeasibility::WillFail;
```

Comment:
I think this could be simplified a bit:

```
if (source->isAnyExistentialType() && isa<MetatypeType>(sourceMetatype) && isa<ExistentialMetatypeType>(targetMetatype))
```

This should also catch cases like `(P & Q).self as P.Type` or `P.Type.Protocol as P.Type.Type`.